src/storage/src/render/upsert.rs

```diff
@@ -8,154 +8,136 @@
 use mz_timely_util::builder_async::{Event as AsyncEvent, OperatorBuilder as AsyncOperatorBuilder};
 
-#[derive(Debug, Clone, PartialOrd, Ord, PartialEq, Eq, Serialize, Deserialize)]
-pub(crate) enum UpsertCommand {
-    Insert(Result<Row, UpsertError>),
-    Remove(Result<Row, UpsertError>),
+#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct UpsertKey([u8; 32]);
+
+thread_local! {
+    pub static KEY_DATUMS: RefCell<DatumVec> = RefCell::new(DatumVec::new());
 }
 
-impl UpsertCommand {
-    fn keyed<'a>(&'a self, key_indices: &'a [usize]) -> KeyedCommand<'a, &Self> {
-        KeyedCommand {
-            command: self,
-            key_indices,
-        }
-    }
+/// The hash function used to map upsert keys. It is important that this hash is a cryptographic
+/// hash so that there is no risk of collisions.
+type KeyHash = Sha256;
 
-    fn into_keyed(self, key_indices: &[usize]) -> KeyedCommand<'_, Self> {
-        KeyedCommand {
-            command: self,
-            key_indices,
-        }
+impl UpsertKey {
+    pub fn from_key(key: Result<&Row, &UpsertError>) -> Self {
+        Self::from_iter(key.map(|r| r.iter()))
     }
-}
 
-/// An UpsertCommand wrapper whose implementations of Hash,PartialEq, and Eq are such that any
-/// variant with the same key is considered the same.
-struct KeyedCommand<'a, C: Borrow<UpsertCommand>> {
-    command: C,
-    key_indices: &'a [usize],
-}
+    pub fn from_value(value: Result<&Row, &UpsertError>, mut key_indices: &[usize]) -> Self {
+        Self::from_iter(value.map(|value| {
+            value.iter().enumerate().flat_map(move |(idx, datum)| {
+                let key_idx = key_indices.get(0)?;
+                if idx == *key_idx {
+                    key_indices = &key_indices[1..];
+                    Some(datum)
+                } else {
+                    None
+                }
+            })
+        }))
+    }
 
-impl<'a, C: Borrow<UpsertCommand>> KeyedCommand<'a, C> {
-    /// Produces the key of this upsert command. If the command has a well-formed key then an
-    /// `Ok(_)` variant is returned with an iterator over the datum keys. If the command is about a
-    /// key error then an `Err(_)` variant is returned.
-    fn key(&self) -> Result<impl Iterator<Item = Datum<'_>> + '_, Datum<'_>> {
-        match self.command.borrow() {
-            UpsertCommand::Insert(Ok(row)) => {
-                let mut key_indices = &*self.key_indices;
-                let iter = row.iter().enumerate().flat_map(move |(idx, datum)| {
-                    let key_idx = key_indices.get(0)?;
-                    if idx == *key_idx {
-                        key_indices = &key_indices[1..];
-                        Some(datum)
-                    } else {
-                        None
+    pub fn from_iter<'a, 'b>(
+        key: Result<impl Iterator<Item = Datum<'a>> + 'b, &UpsertError>,
+    ) -> Self {
```

Comment:
can we just pass a shared datumvec in from the operator state instead of using the thread local?